emacs/.doom.d/config.org
#+TITLE: Doom Emacs Configuration

This is an ongoing evolution of my original Emacs configuration files, inspired by a bunch of resources I've found online.

* Table of Contents :TOC:
- [[#doom-configuration-functions-and-macros][Doom Configuration Functions and Macros]]
- [[#preamble][Preamble]]
- [[#personal-info][Personal Info]]
- [[#appearance][Appearance]]
  - [[#font][Font]]
  - [[#theme][Theme]]
  - [[#line-numbers][Line Numbers]]
  - [[#global-key-maps][Global key maps]]
- [[#emacs-customize][Emacs Customize]]
- [[#projectile][Projectile]]
- [[#org-mode][Org mode]]
- [[#lsp][LSP]]
- [[#ccpp][C/CPP]]
  - [[#makefile-generation][Makefile generation]]
  - [[#code-indexing][Code indexing]]
  - [[#debugging][Debugging]]

* Doom Configuration Functions and Macros

Here are some additional functions/macros that could help you configure Doom:

+ =load!= for loading external =*.el= files relative to this one
+ =use-package!= for configuring packages
+ =after!= for running code after a package has loaded
+ =add-load-path!' for adding directories to the =load-path=, relative to this file. Emacs searches the =load-path= when you load packages with =require= or =use-package=.
+ =map!= for binding new keys

* Preamble

#+begin_src emacs-lisp :tangle yes
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
#+end_src

* Personal Info

Some functionality uses this to identify you, e.g. GPG configuration, email clients, file templates and snippets.

#+begin_src emacs-lisp :tangle yes
(setq user-full-name "Donald Thompson"
      user-mail-address "donald@witt3rd.com")
#+end_src

* Appearance

This section affects the themes, fonts, and other visual elements.

** Font

Doom exposes five (optional) variables for controlling fonts in Doom. Here are the three important ones:

+ =doom-font=
+ =doom-variable-pitch-font=
+ =doom-big-font= -- used for =doom-big-font-mode=; use this for presentations or streaming.

They all accept either a font-spec, font string (="Input Mono-12"=), or xlfd font string. You generally only need these two:

#+begin_src emacs-lisp :tangle yes
(setq doom-font (font-spec :family "JetBrains Mono" :size 16 :weight 'regular)
      doom-variable-pitch-font (font-spec :family "Cantarell" :size 18)
      doom-big-font (font-spec :family "Cantarell" :size 36 :weight 'bold))
#+end_src

** Theme

 There are two ways to load a theme. Both assume the theme is installed and available. You can either set =doom-theme= or manually load a theme with the =load-theme= function. This is the default:

#+begin_src emacs-lisp :tangle yes
(setq custom-safe-themes t)
(if (window-system)
    (setq doom-theme 'doom-dark+)
  (setq doom-theme 'doom-one))
(setq fancy-splash-image "~/.doom.d/doom.png")
#+end_src

** Line Numbers

This determines the style of line numbers in effect. If set to =nil=, line numbers are disabled. For relative line numbers, set this to =relative=.

#+begin_src emacs-lisp :tangle yes
(setq display-line-numbers-type 'relative)
#+end_src

#+RESULTS:
: relative

** Global key maps

Define custom keys for indenting regions:

#+begin_src emacs-lisp :tangle yes
(map! "C->" 'indent-rigidly-right-to-tab-stop)
(map! "C-<" 'indent-rigidly-left-to-tab-stop)
#+end_src

* Emacs Customize

Ignore emacs "customize" by redirecting the file to /dev/null

#+begin_src emacs-lisp :tangle yes
(setq custom-file null-device)
#+end_src

* Projectile

Set project search paths and settings

#+begin_src emacs-lisp :tangle yes
(setq projectile-project-search-path '(("~/src/witt3rd/" . 5) "~/org" ("~/dotfiles" . 0))
      projectile-auto-discover 1
 )
#+end_src

* Org mode

Set the root directory.  It must be set before org loads!

#+begin_src emacs-lisp :tangle yes
(after! org

  (setq org-directory "~/org/"
        org-ellipsis " ▼"
        org-hide-emphasis-markers nil
        org-agenda-files (list "~/org")
        org-log-done 'time
        org-log-into-drawer t
        )

; (defun my/org-mode-setup ()
;    (org-indent-mode 0)
;    (variable-pitch-mode 0)
;    (auto-fill-mode 0)
;    (visual-line-mode 1)
;    (dolist (face '((org-level-1 . 1.2)
;                    (org-level-2 . 1.1)
;                    (org-level-3 . 1.05)
;                    (org-level-4 . 1.0)
;                    (org-level-5 . 1.0)
;                    (org-level-6 . 1.0)
;                    (org-level-7 . 1.0)
;                    (org-level-8 . 1.0)))
;      (set-face-attribute (car face) nil :font "Cantarell" :weight 'bold :height (cdr face)))
;    (set-face-attribute 'org-block nil :foreground nil :inherit 'fixed-pitch)
;    (set-face-attribute 'org-code nil :inherit '(shadow fixed-pitch))
;    (set-face-attribute 'org-indent nil :inherit '(org-hide fixed-pitch))
;    (set-face-attribute 'org-verbatim nil :inherit '(shadow fixed-pitch))
;    (set-face-attribute 'org-special-keyword nil :inherit '(font-lock-comment-face fixed-pitch))
;    (set-face-attribute 'org-meta-line nil :inherit '(font-lock-comment-face fixed-pitch))
;    (set-face-attribute 'org-checkbox nil :inherit 'fixed-pitch)
;    )
;
;  (add-hook 'org-mode-hook #'my/org-mode-setup)
  )
#+end_src

Center the org-mode document:

#+begin_src emacs-lisp :tangle yes
;(use-package! visual-fill-column
;  :after org
;  :config
;  (defun my/org-mode-visual-fill ()
;    (setq visual-fill-column-width 200
;          visual-fill-column-center-text t)
;    (visual-fill-column-mode 1)
;    )
;
;  (add-hook 'org-mode-hook #'my/org-mode-visual-fill)
;  )
#+end_src

Add [[https://github.com/alphapapa/org-super-agenda][org-super-agenda]] for nicer agenda views

#+begin_src emacs-lisp :tangle yes
(use-package! org-super-agenda
  :after org-agenda
  :init
  (setq org-super-agenda-groups '((:name "Today"
                                   :time-grid t
                                   :scheduled today)
                                  (:name "Due Today"
                                   :deadline today)
                                  (:name "Important"
                                   :priority "A")
                                  (:name "Overdue"
                                   :deadline past)
                                  (:name "Due soon"
                                   :deadline future)
                                  (:name "Big Outcomes"
                                   :tag "bo")))
  :config
  (org-super-agenda-mode)
  )
#+end_src

* LSP

#+begin_src emacs-lisp :tangle yes
(setq lsp-clients-clangd-args '("-j=3"
                                "--background-index"
                                "--clang-tidy"
                                "--completion-style=detailed"
                                "--header-insertion=iwyu"
                                "--header-insertion-decorators=0"))
(setq lsp-ui-mode nil)
(after! lsp-clangd (set-lsp-priority! 'clangd 2))
#+end_src

* C/CPP

** Makefile generation

[[https://github.com/atilaneves/cmake-ide][cmake-ide]]

#+begin_src emacs-lisp :tangle yes
(after! projectile

  (defun my/cmake-ide-find-project ()
    "Finds the directory of the project for cmake-ide."
    (with-eval-after-load 'projectile
      (setq cmake-ide-project-dir (projectile-project-root))
      (setq cmake-ide-build-dir (concat cmake-ide-project-dir "build")))
    (setq cmake-ide-compile-command
          (concat "cd " cmake-ide-build-dir " && cmake .. && make"))
    (cmake-ide-load-db))

  (defun my/switch-to-compilation-window ()
    "Switches to the *compilation* buffer after compilation."
    (other-window 1))

  ;; TODO
  ;;:bind ([remap comment-region] . cmake-ide-compile)

  (advice-add 'cmake-ide-compile :after #'my/switch-to-compilation-window)

  (add-hook 'c++-mode-hook #'my/cmake-ide-find-project)
)
#+end_src

** Code indexing

[[https://github.com/Andersbakken/rtags][RTags]]

#+begin_src emacs-lisp :tangle yes
(require 'rtags)
#+end_src

** Debugging

[[https://github.com/emacs-lsp/dap-mode/blob/master/dap-cpptools.el][dap-cpptools]]

#+begin_src emacs-lisp :tangle yes
(require 'dap-cpptools)
#+end_src


emacs/.doom.d/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

(setq user-full-name "Donald Thompson"
      user-mail-address "donald@witt3rd.com")

(setq doom-font (font-spec :family "JetBrains Mono" :size 16 :weight 'regular)
      doom-variable-pitch-font (font-spec :family "Cantarell" :size 18)
      doom-big-font (font-spec :family "Cantarell" :size 36 :weight 'bold))

(setq custom-safe-themes t)
(if (window-system)
    (setq doom-theme 'doom-dark+)
  (setq doom-theme 'doom-one))
(setq fancy-splash-image "~/.doom.d/doom.png")

(setq display-line-numbers-type 'relative)

(map! "C->" 'indent-rigidly-right-to-tab-stop)
(map! "C-<" 'indent-rigidly-left-to-tab-stop)

(setq custom-file null-device)

(setq projectile-project-search-path '(("~/src/witt3rd/" . 5) "~/org" ("~/dotfiles" . 0))
      projectile-auto-discover 1
 )

(after! org

  (setq org-directory "~/org/"
        org-ellipsis " ▼"
        org-hide-emphasis-markers nil
        org-agenda-files (list "~/org")
        org-log-done 'time
        org-log-into-drawer t
        )

; (defun my/org-mode-setup ()
;    (org-indent-mode 0)
;    (variable-pitch-mode 0)
;    (auto-fill-mode 0)
;    (visual-line-mode 1)
;    (dolist (face '((org-level-1 . 1.2)
;                    (org-level-2 . 1.1)
;                    (org-level-3 . 1.05)
;                    (org-level-4 . 1.0)
;                    (org-level-5 . 1.0)
;                    (org-level-6 . 1.0)
;                    (org-level-7 . 1.0)
;                    (org-level-8 . 1.0)))
;      (set-face-attribute (car face) nil :font "Cantarell" :weight 'bold :height (cdr face)))
;    (set-face-attribute 'org-block nil :foreground nil :inherit 'fixed-pitch)
;    (set-face-attribute 'org-code nil :inherit '(shadow fixed-pitch))
;    (set-face-attribute 'org-indent nil :inherit '(org-hide fixed-pitch))
;    (set-face-attribute 'org-verbatim nil :inherit '(shadow fixed-pitch))
;    (set-face-attribute 'org-special-keyword nil :inherit '(font-lock-comment-face fixed-pitch))
;    (set-face-attribute 'org-meta-line nil :inherit '(font-lock-comment-face fixed-pitch))
;    (set-face-attribute 'org-checkbox nil :inherit 'fixed-pitch)
;    )
;
;  (add-hook 'org-mode-hook #'my/org-mode-setup)
  )

;(use-package! visual-fill-column
;  :after org
;  :config
;  (defun my/org-mode-visual-fill ()
;    (setq visual-fill-column-width 200
;          visual-fill-column-center-text t)
;    (visual-fill-column-mode 1)
;    )
;
;  (add-hook 'org-mode-hook #'my/org-mode-visual-fill)
;  )

(use-package! org-super-agenda
  :after org-agenda
  :init
  (setq org-super-agenda-groups '((:name "Today"
                                   :time-grid t
                                   :scheduled today)
                                  (:name "Due Today"
                                   :deadline today)
                                  (:name "Important"
                                   :priority "A")
                                  (:name "Overdue"
                                   :deadline past)
                                  (:name "Due soon"
                                   :deadline future)
                                  (:name "Big Outcomes"
                                   :tag "bo")))
  :config
  (org-super-agenda-mode)
  )

(setq lsp-clients-clangd-args '("-j=3"
                                "--background-index"
                                "--clang-tidy"
                                "--completion-style=detailed"
                                "--header-insertion=iwyu"
                                "--header-insertion-decorators=0"))
(setq lsp-ui-mode nil)
(after! lsp-clangd (set-lsp-priority! 'clangd 2))

(after! projectile

  (defun my/cmake-ide-find-project ()
    "Finds the directory of the project for cmake-ide."
    (with-eval-after-load 'projectile
      (setq cmake-ide-project-dir (projectile-project-root))
      (setq cmake-ide-build-dir (concat cmake-ide-project-dir "build")))
    (setq cmake-ide-compile-command
          (concat "cd " cmake-ide-build-dir " && cmake .. && make"))
    (cmake-ide-load-db))

  (defun my/switch-to-compilation-window ()
    "Switches to the *compilation* buffer after compilation."
    (other-window 1))

  ;; TODO
  ;;:bind ([remap comment-region] . cmake-ide-compile)

  (advice-add 'cmake-ide-compile :after #'my/switch-to-compilation-window)

  (add-hook 'c++-mode-hook #'my/cmake-ide-find-project)
)

(require 'rtags)

(require 'dap-cpptools)
